File: app/src/main/java/com/flynes/emu/catalog/scan/RomPayloadParser.java
package com.flynes.emu.catalog.scan;

import com.flynes.emu.catalog.CompatibilityDecision;
import com.flynes.emu.catalog.CompatibilityReason;
import com.flynes.emu.catalog.CompatibilityState;
import com.flynes.emu.catalog.RomAnalysis;
import com.flynes.emu.catalog.RomFormat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class RomPayloadParser {
    private static final int INES_HEADER_BYTES = 16;
    private static final int TRAINER_BYTES = 512;
    private static final int FDS_HEADER_BYTES = 16;
    private static final int FDS_SIDE_BYTES = 65_500;
    private static final byte[] FDS_SIDE_SIGNATURE = new byte[]{
            1, '*', 'N', 'I', 'N', 'T', 'E', 'N', 'D', 'O', '-', 'H', 'V', 'C', '*'};

    private RomPayloadParser() {
    }

    static Parsed parse(byte[] payload) {
        if (startsWith(payload, new byte[]{'N', 'E', 'S', 0x1A})) {
            return parseNes(payload);
        }
        if (startsWith(payload, new byte[]{'F', 'D', 'S', 0x1A})) {
            return parseHeaderedFds(payload);
        }
        if (looksLikeHeaderlessFds(payload)) {
            return parseHeaderlessFds(payload);
        }
        if (startsWith(payload, new byte[]{'U', 'N', 'I', 'F'})) {
            return parseUnif(payload);
        }
        return null;
    }

    private static Parsed parseNes(byte[] payload) {
        if (payload.length < INES_HEADER_BYTES) {
            return invalid(
                    RomFormat.INES,
                    CompatibilityReason.NES_HEADER_INVALID,
                    RomAnalysis.basic(payload.length));
        }
        int flags6 = unsigned(payload[6]);
        int flags7 = unsigned(payload[7]);
        boolean nes2 = (flags7 & 0x0C) == 0x08;
        RomFormat format = nes2 ? RomFormat.NES2 : RomFormat.INES;
        boolean trainer = (flags6 & 0x04) != 0;
        boolean battery = (flags6 & 0x02) != 0;
        int mapper = (flags6 >>> 4) | (flags7 & 0xF0);
        int submapper = 0;
        long prgBytes;
        long chrBytes;
        try {
            if (nes2) {
                int byte8 = unsigned(payload[8]);
                int byte9 = unsigned(payload[9]);
                mapper |= (byte8 & 0x0F) << 8;
                submapper = byte8 >>> 4;
                prgBytes = nes2Size(unsigned(payload[4]), byte9 & 0x0F, 16_384L);
                chrBytes = nes2Size(unsigned(payload[5]), byte9 >>> 4, 8_192L);
            } else {
                prgBytes = Math.multiplyExact((long) unsigned(payload[4]), 16_384L);
                chrBytes = Math.multiplyExact((long) unsigned(payload[5]), 8_192L);
            }
        } catch (ArithmeticException overflow) {
            return invalid(
                    format,
                    CompatibilityReason.NES_SIZE_OVERFLOW,
                    new RomAnalysis(
                            0, payload.length, 0, 0, mapper, submapper,
                            trainer, battery, 0, Collections.<RomAnalysis.Warning>emptyList()));
        }

        ArrayList<RomAnalysis.Warning> warnings = new ArrayList<>();
        if (!nes2 && hasDirtyLegacyHeader(payload)) {
            warnings.add(RomAnalysis.Warning.DIRTY_HEADER);
        }
        long expected;
        try {
            expected = Math.addExact(
                    INES_HEADER_BYTES + (trainer ? TRAINER_BYTES : 0L),
                    Math.addExact(prgBytes, chrBytes));
        } catch (ArithmeticException overflow) {
            return invalid(
                    format,
                    CompatibilityReason.NES_SIZE_OVERFLOW,
                    new RomAnalysis(
                            0, payload.length, prgBytes, chrBytes, mapper, submapper,
                            trainer, battery, 0, warnings));
        }
        if (prgBytes == 0) {
            return invalid(
                    format,
                    CompatibilityReason.NES_ZERO_PRG,
                    new RomAnalysis(
                            expected, payload.length, prgBytes, chrBytes, mapper, submapper,
                            trainer, battery, 0, warnings));
        }
        if (payload.length < expected) {
            return invalid(
                    format,
                    CompatibilityReason.NES_TRUNCATED,
                    new RomAnalysis(
                            expected, payload.length, prgBytes, chrBytes, mapper, submapper,
                            trainer, battery, 0, warnings));
        }
        if (payload.length > expected) {
            warnings.add(RomAnalysis.Warning.TRAILING_DATA);
        }
        return new Parsed(
                format,
                CompatibilityDecision.playableNes(),
                new RomAnalysis(
                        expected, payload.length, prgBytes, chrBytes, mapper, submapper,
                        trainer, battery, 0, warnings));
    }

    private static long nes2Size(int lsb, int msbNibble, long unit) {
        if (msbNibble != 0x0F) {
            return Math.multiplyExact(((long) msbNibble << 8) | lsb, unit);
        }
        int exponent = lsb >>> 2;
        int multiplier = ((lsb & 0x03) * 2) + 1;
        if (exponent >= 63) {
            throw new ArithmeticException("NES2 exponential size overflows signed long");
        }
        long base = 1L << exponent;
        return Math.multiplyExact(base, (long) multiplier);
    }

    private static Parsed parseHeaderedFds(byte[] payload) {
        if (payload.length < FDS_HEADER_BYTES) {
            return invalid(
                    RomFormat.FDS,
                    CompatibilityReason.FDS_TRUNCATED,
                    RomAnalysis.basic(payload.length));
        }
        int sides = unsigned(payload[4]);
        if (sides == 0) {
            return invalid(
                    RomFormat.FDS,
                    CompatibilityReason.FDS_INVALID_SIDE_COUNT,
                    RomAnalysis.basic(payload.length));
        }
        long expected = FDS_HEADER_BYTES + sides * (long) FDS_SIDE_BYTES;
        if (payload.length < expected) {
            return invalid(
                    RomFormat.FDS,
                    CompatibilityReason.FDS_TRUNCATED,
                    fdsAnalysis(expected, payload.length, sides, Collections.<RomAnalysis.Warning>emptyList()));
        }
        if (!validFdsSignatures(payload, FDS_HEADER_BYTES, sides)) {
            return invalid(
                    RomFormat.FDS,
                    CompatibilityReason.FDS_INVALID_HEADER,
                    fdsAnalysis(expected, payload.length, sides, Collections.<RomAnalysis.Warning>emptyList()));
        }
        ArrayList<RomAnalysis.Warning> warnings = new ArrayList<>();
        if (payload.length > expected) {
            warnings.add(RomAnalysis.Warning.TRAILING_DATA);
        }
        return unsupportedFds(fdsAnalysis(expected, payload.length, sides, warnings));
    }

    private static Parsed parseHeaderlessFds(byte[] payload) {
        if (payload.length < FDS_SIDE_BYTES || payload.length % FDS_SIDE_BYTES != 0) {
            int expectedSides = (int) Math.max(
                    1L, (payload.length + (long) FDS_SIDE_BYTES - 1L) / FDS_SIDE_BYTES);
            long expected = expectedSides * (long) FDS_SIDE_BYTES;
            return invalid(
                    RomFormat.FDS,
                    CompatibilityReason.FDS_TRUNCATED,
                    fdsAnalysis(expected, payload.length, expectedSides,
                            Collections.<RomAnalysis.Warning>emptyList()));
        }
        int sides = payload.length / FDS_SIDE_BYTES;
        if (!validFdsSignatures(payload, 0, sides)) {
            return invalid(
                    RomFormat.FDS,
                    CompatibilityReason.FDS_INVALID_HEADER,
                    fdsAnalysis(payload.length, payload.length, sides,
                            Collections.<RomAnalysis.Warning>emptyList()));
        }
        return unsupportedFds(fdsAnalysis(
                payload.length, payload.length, sides,
                Collections.<RomAnalysis.Warning>emptyList()));
    }

    private static Parsed unsupportedFds(RomAnalysis analysis) {
        return new Parsed(
                RomFormat.FDS,
                new CompatibilityDecision(
                        CompatibilityState.UNSUPPORTED,
                        CompatibilityReason.FDS_BIOS_API_NOT_IMPLEMENTED),
                analysis);
    }

    private static RomAnalysis fdsAnalysis(
            long expected, long actual, int sides, List<RomAnalysis.Warning> warnings) {
        return new RomAnalysis(
                expected, actual, 0, 0, -1, -1,
                false, false, sides, warnings);
    }

    private static boolean looksLikeHeaderlessFds(byte[] payload) {
        return startsWithAt(payload, 0, FDS_SIDE_SIGNATURE);
    }

    private static boolean validFdsSignatures(byte[] payload, int offset, int sides) {
        for (int side = 0; side < sides; side++) {
            if (!startsWithAt(payload, offset + side * FDS_SIDE_BYTES, FDS_SIDE_SIGNATURE)) {
                return false;
            }
        }
        return true;
    }

    private static Parsed parseUnif(byte[] payload) {
        if (payload.length < 32) {
            return invalid(
                    RomFormat.UNIF,
                    CompatibilityReason.UNIF_INVALID_CHUNK,
                    RomAnalysis.basic(payload.length));
        }
        int cursor = 32;
        boolean hasPrg = false;
        while (cursor < payload.length) {
            if (payload.length - cursor < 8) {
                return invalid(
                        RomFormat.UNIF,
                        CompatibilityReason.UNIF_INVALID_CHUNK,
                        RomAnalysis.basic(payload.length));
            }
            boolean prgChunk = payload[cursor] == 'P'
                    && payload[cursor + 1] == 'R'
                    && payload[cursor + 2] == 'G';
            long length = unsignedInt(payload, cursor + 4);
            cursor += 8;
            if (length > payload.length - cursor) {
                return invalid(
                        RomFormat.UNIF,
                        CompatibilityReason.UNIF_INVALID_CHUNK,
                        RomAnalysis.basic(payload.length));
            }
            if (prgChunk && length > 0) {
                hasPrg = true;
            }
            cursor += (int) length;
        }
        if (!hasPrg) {
            return invalid(
                    RomFormat.UNIF,
                    CompatibilityReason.UNIF_MISSING_PRG,
                    RomAnalysis.basic(payload.length));
        }
        return new Parsed(
                RomFormat.UNIF,
                new CompatibilityDecision(
                        CompatibilityState.UNSUPPORTED,
                        CompatibilityReason.UNIF_PRODUCT_DISABLED),
                RomAnalysis.basic(payload.length));
    }

    private static Parsed invalid(
            RomFormat format, CompatibilityReason reason, RomAnalysis analysis) {
        return new Parsed(
                format,
                new CompatibilityDecision(CompatibilityState.INVALID, reason),
                analysis);
    }

    private static boolean hasDirtyLegacyHeader(byte[] payload) {
        for (int index = 12; index < 16; index++) {
            if (payload[index] != 0) {
                return true;
            }
        }
        return false;
    }

    private static boolean startsWith(byte[] value, byte[] prefix) {
        return startsWithAt(value, 0, prefix);
    }

    private static boolean startsWithAt(byte[] value, int offset, byte[] prefix) {
        if (offset < 0 || value.length - offset < prefix.length) {
            return false;
        }
        for (int index = 0; index < prefix.length; index++) {
            if (value[offset + index] != prefix[index]) {
                return false;
            }
        }
        return true;
    }

    private static int unsigned(byte value) {
        return value & 0xFF;
    }

    private static long unsignedInt(byte[] value, int offset) {
        return (value[offset] & 0xFFL)
                | ((value[offset + 1] & 0xFFL) << 8)
                | ((value[offset + 2] & 0xFFL) << 16)
                | ((value[offset + 3] & 0xFFL) << 24);
    }

    record Parsed(
            RomFormat format,
            CompatibilityDecision compatibility,
            RomAnalysis analysis) {
    }
}

File: shared/src/catalog/rom_payload_parser.cpp
#include "catalog/rom_payload_parser.hpp"

#include <array>
#include <cstddef>
#include <cstdint>
#include <limits>

namespace flynes::catalog {
namespace {

constexpr std::size_t ines_header_bytes = 16u;
constexpr std::uint64_t trainer_bytes = 512u;
constexpr std::size_t fds_header_bytes = 16u;
constexpr std::size_t fds_side_bytes = 65'500u;
constexpr std::uint64_t signed_64_max =
    static_cast<std::uint64_t>(std::numeric_limits<std::int64_t>::max());

constexpr std::array<std::uint8_t, 4> nes_signature = {
    static_cast<std::uint8_t>('N'),
    static_cast<std::uint8_t>('E'),
    static_cast<std::uint8_t>('S'),
    0x1Au,
};
constexpr std::array<std::uint8_t, 4> fds_signature = {
    static_cast<std::uint8_t>('F'),
    static_cast<std::uint8_t>('D'),
    static_cast<std::uint8_t>('S'),
    0x1Au,
};
constexpr std::array<std::uint8_t, 4> unif_signature = {
    static_cast<std::uint8_t>('U'),
    static_cast<std::uint8_t>('N'),
    static_cast<std::uint8_t>('I'),
    static_cast<std::uint8_t>('F'),
};
constexpr std::array<std::uint8_t, 15> fds_side_signature = {
    0x01u,
    static_cast<std::uint8_t>('*'),
    static_cast<std::uint8_t>('N'),
    static_cast<std::uint8_t>('I'),
    static_cast<std::uint8_t>('N'),
    static_cast<std::uint8_t>('T'),
    static_cast<std::uint8_t>('E'),
    static_cast<std::uint8_t>('N'),
    static_cast<std::uint8_t>('D'),
    static_cast<std::uint8_t>('O'),
    static_cast<std::uint8_t>('-'),
    static_cast<std::uint8_t>('H'),
    static_cast<std::uint8_t>('V'),
    static_cast<std::uint8_t>('C'),
    static_cast<std::uint8_t>('*'),
};

static_assert(sizeof(std::size_t) <= sizeof(std::uint64_t),
              "ROM analysis requires size_t to fit in uint64_t");

std::uint64_t byte_count(std::size_t size) noexcept
{
    return static_cast<std::uint64_t>(size);
}

template <std::size_t Size>
bool starts_with_at(ByteView payload,
                    std::size_t offset,
                    const std::array<std::uint8_t, Size>& prefix) noexcept
{
    if (payload.data == nullptr || offset > payload.size || payload.size - offset < Size)
    {
        return false;
    }
    for (std::size_t index = 0; index < Size; ++index)
    {
        if (payload.data[offset + index] != prefix[index])
        {
            return false;
        }
    }
    return true;
}

RomAnalysis basic_analysis(std::size_t actual_size) noexcept
{
    RomAnalysis analysis;
    analysis.expected_bytes = byte_count(actual_size);
    analysis.actual_bytes = byte_count(actual_size);
    return analysis;
}

RomParseResult unrecognized(std::size_t actual_size) noexcept
{
    RomParseResult result;
    result.analysis = basic_analysis(actual_size);
    return result;
}

RomParseResult invalid(RomFormat format,
                       CompatibilityReason reason,
                       const RomAnalysis& analysis) noexcept
{
    RomParseResult result;
    result.recognized = true;
    result.format = format;
    result.state = CompatibilityState::INVALID;
    result.reason = reason;
    result.analysis = analysis;
    return result;
}

void append_warning(RomAnalysis& analysis, RomWarning warning) noexcept
{
    if (analysis.warning_count < analysis.warnings.size())
    {
        analysis.warnings[analysis.warning_count] = warning;
        ++analysis.warning_count;
    }
}

bool checked_signed_add(std::uint64_t left,
                        std::uint64_t right,
                        std::uint64_t& result) noexcept
{
    if (left > signed_64_max || right > signed_64_max || left > signed_64_max - right)
    {
        return false;
    }
    result = left + right;
    return true;
}

bool checked_signed_multiply(std::uint64_t left,
                             std::uint64_t right,
                             std::uint64_t& result) noexcept
{
    if (left != 0u && right > signed_64_max / left)
    {
        return false;
    }
    result = left * right;
    return true;
}

bool nes2_size(std::uint8_t low,
               std::uint8_t high_nibble,
               std::uint64_t unit,
               std::uint64_t& result) noexcept
{
    if (high_nibble != 0x0Fu)
    {
        const std::uint64_t count =
            (static_cast<std::uint64_t>(high_nibble) << 8u) |
            static_cast<std::uint64_t>(low);
        return checked_signed_multiply(count, unit, result);
    }

    const std::uint8_t exponent = static_cast<std::uint8_t>(low >> 2u);
    const std::uint8_t multiplier =
        static_cast<std::uint8_t>(((low & 0x03u) * 2u) + 1u);
    if (exponent >= 63u)
    {
        return false;
    }
    const std::uint64_t base = std::uint64_t{1} << exponent;
    return checked_signed_multiply(base, multiplier, result);
}

bool has_dirty_legacy_header(ByteView payload) noexcept
{
    for (std::size_t index = 12u; index < ines_header_bytes; ++index)
    {
        if (payload.data[index] != 0u)
        {
            return true;
        }
    }
    return false;
}

RomParseResult parse_nes(ByteView payload) noexcept
{
    if (payload.size < ines_header_bytes)
    {
        return invalid(RomFormat::INES,
                       CompatibilityReason::NES_HEADER_INVALID,
                       basic_analysis(payload.size));
    }

    const std::uint32_t flags6 = payload.data[6];
    const std::uint32_t flags7 = payload.data[7];
    const bool is_nes2 = (flags7 & 0x0Cu) == 0x08u;
    const RomFormat format = is_nes2 ? RomFormat::NES2 : RomFormat::INES;
    const bool trainer = (flags6 & 0x04u) != 0u;
    const bool battery = (flags6 & 0x02u) != 0u;
    std::int32_t mapper =
        static_cast<std::int32_t>((flags6 >> 4u) | (flags7 & 0xF0u));
    std::int32_t submapper = 0;
    std::uint64_t prg_bytes = 0;
    std::uint64_t chr_bytes = 0;

    if (is_nes2)
    {
        const std::uint32_t byte8 = payload.data[8];
        const std::uint8_t byte9 = payload.data[9];
        mapper = static_cast<std::int32_t>(
            static_cast<std::uint32_t>(mapper) | ((byte8 & 0x0Fu) << 8u));
        submapper = static_cast<std::int32_t>(byte8 >> 4u);
        if (!nes2_size(payload.data[4],
                       static_cast<std::uint8_t>(byte9 & 0x0Fu),
                       16'384u,
                       prg_bytes) ||
            !nes2_size(payload.data[5],
                       static_cast<std::uint8_t>(byte9 >> 4u),
                       8'192u,
                       chr_bytes))
        {
            RomAnalysis analysis;
            analysis.actual_bytes = byte_count(payload.size);
            analysis.mapper = mapper;
            analysis.submapper = submapper;
            analysis.trainer = trainer;
            analysis.battery = battery;
            return invalid(format, CompatibilityReason::NES_SIZE_OVERFLOW, analysis);
        }
    }
    else
    {
        const bool valid_prg = checked_signed_multiply(payload.data[4], 16'384u, prg_bytes);
        const bool valid_chr = checked_signed_multiply(payload.data[5], 8'192u, chr_bytes);
        if (!valid_prg || !valid_chr)
        {
            RomAnalysis analysis;
            analysis.actual_bytes = byte_count(payload.size);
            analysis.mapper = mapper;
            analysis.submapper = submapper;
            analysis.trainer = trainer;
            analysis.battery = battery;
            return invalid(format, CompatibilityReason::NES_SIZE_OVERFLOW, analysis);
        }
    }

    RomAnalysis analysis;
    analysis.actual_bytes = byte_count(payload.size);
    analysis.prg_bytes = prg_bytes;
    analysis.chr_bytes = chr_bytes;
    analysis.mapper = mapper;
    analysis.submapper = submapper;
    analysis.trainer = trainer;
    analysis.battery = battery;
    if (!is_nes2 && has_dirty_legacy_header(payload))
    {
        append_warning(analysis, RomWarning::DIRTY_HEADER);
        analysis.mapper = static_cast<std::int32_t>(flags6 >> 4u);
    }

    std::uint64_t content_bytes = 0;
    const std::uint64_t prefix_bytes =
        byte_count(ines_header_bytes) + (trainer ? trainer_bytes : 0u);
    if (!checked_signed_add(prg_bytes, chr_bytes, content_bytes) ||
        !checked_signed_add(prefix_bytes, content_bytes, analysis.expected_bytes))
    {
        analysis.expected_bytes = 0;
        return invalid(format, CompatibilityReason::NES_SIZE_OVERFLOW, analysis);
    }
    if (prg_bytes == 0u)
    {
        return invalid(format, CompatibilityReason::NES_ZERO_PRG, analysis);
    }
    if (analysis.actual_bytes < analysis.expected_bytes)
    {
        return invalid(format, CompatibilityReason::NES_TRUNCATED, analysis);
    }
    if (analysis.actual_bytes > analysis.expected_bytes)
    {
        append_warning(analysis, RomWarning::TRAILING_DATA);
    }

    RomParseResult result;
    result.recognized = true;
    result.format = format;
    result.state = CompatibilityState::PLAYABLE;
    result.reason = CompatibilityReason::PLAYABLE_NES;
    result.analysis = analysis;
    return result;
}

RomAnalysis fds_analysis(std::uint64_t expected,
                         std::size_t actual,
                         std::uint64_t sides) noexcept
{
    RomAnalysis analysis;
    analysis.expected_bytes = expected;
    analysis.actual_bytes = byte_count(actual);
    analysis.disk_sides = sides;
    return analysis;
}

bool valid_fds_signatures(ByteView payload,
                          std::size_t offset,
                          std::size_t sides) noexcept
{
    std::size_t side_offset = offset;
    for (std::size_t side = 0; side < sides; ++side)
    {
        if (!starts_with_at(payload, side_offset, fds_side_signature))
        {
            return false;
        }
        if (side + 1u < sides)
        {
            side_offset += fds_side_bytes;
        }
    }
    return true;
}

RomParseResult unsupported_fds(const RomAnalysis& analysis) noexcept
{
    RomParseResult result;
    result.recognized = true;
    result.format = RomFormat::FDS;
    result.state = CompatibilityState::UNSUPPORTED;
    result.reason = CompatibilityReason::FDS_BIOS_API_NOT_IMPLEMENTED;
    result.analysis = analysis;
    return result;
}

RomParseResult parse_headered_fds(ByteView payload) noexcept
{
    if (payload.size < fds_header_bytes)
    {
        return invalid(RomFormat::FDS,
                       CompatibilityReason::FDS_TRUNCATED,
                       basic_analysis(payload.size));
    }
    const std::uint8_t sides = payload.data[4];
    if (sides == 0u)
    {
        return invalid(RomFormat::FDS,
                       CompatibilityReason::FDS_INVALID_SIDE_COUNT,
                       basic_analysis(payload.size));
    }

    const std::uint64_t expected =
        byte_count(fds_header_bytes) +
        static_cast<std::uint64_t>(sides) * byte_count(fds_side_bytes);
    RomAnalysis analysis = fds_analysis(expected, payload.size, sides);
    if (analysis.actual_bytes < expected)
    {
        return invalid(RomFormat::FDS, CompatibilityReason::FDS_TRUNCATED, analysis);
    }
    if (!valid_fds_signatures(payload, fds_header_bytes, sides))
    {
        return invalid(RomFormat::FDS, CompatibilityReason::FDS_INVALID_HEADER, analysis);
    }
    if (analysis.actual_bytes > expected)
    {
        append_warning(analysis, RomWarning::TRAILING_DATA);
    }
    return unsupported_fds(analysis);
}

RomParseResult parse_headerless_fds(ByteView payload) noexcept
{
    if (payload.size < fds_side_bytes || payload.size % fds_side_bytes != 0u)
    {
        std::size_t expected_sides = payload.size / fds_side_bytes;
        if (payload.size % fds_side_bytes != 0u)
        {
            ++expected_sides;
        }
        if (expected_sides == 0u)
        {
            expected_sides = 1u;
        }
        const std::uint64_t sides = byte_count(expected_sides);
        const std::uint64_t side_bytes = byte_count(fds_side_bytes);
        const std::uint64_t expected =
            sides > std::numeric_limits<std::uint64_t>::max() / side_bytes
                ? std::numeric_limits<std::uint64_t>::max()
                : sides * side_bytes;
        return invalid(
            RomFormat::FDS,
            CompatibilityReason::FDS_TRUNCATED,
            fds_analysis(expected, payload.size, sides));
    }

    const std::size_t sides = payload.size / fds_side_bytes;
    RomAnalysis analysis =
        fds_analysis(byte_count(payload.size), payload.size, byte_count(sides));
    if (!valid_fds_signatures(payload, 0u, sides))
    {
        return invalid(RomFormat::FDS, CompatibilityReason::FDS_INVALID_HEADER, analysis);
    }
    return unsupported_fds(analysis);
}

bool is_hex_digit(std::uint8_t value) noexcept
{
    return (value >= static_cast<std::uint8_t>('0') &&
            value <= static_cast<std::uint8_t>('9')) ||
           (value >= static_cast<std::uint8_t>('A') &&
            value <= static_cast<std::uint8_t>('F')) ||
           (value >= static_cast<std::uint8_t>('a') &&
            value <= static_cast<std::uint8_t>('f'));
}

std::uint64_t little_endian_u32(ByteView payload, std::size_t offset) noexcept
{
    return static_cast<std::uint64_t>(payload.data[offset]) |
           (static_cast<std::uint64_t>(payload.data[offset + 1u]) << 8u) |
           (static_cast<std::uint64_t>(payload.data[offset + 2u]) << 16u) |
           (static_cast<std::uint64_t>(payload.data[offset + 3u]) << 24u);
}

RomParseResult parse_unif(ByteView payload) noexcept
{
    if (payload.size < 32u)
    {
        return invalid(RomFormat::UNIF,
                       CompatibilityReason::UNIF_INVALID_CHUNK,
                       basic_analysis(payload.size));
    }

    std::size_t cursor = 32u;
    bool has_prg = false;
    while (cursor < payload.size)
    {
        if (payload.size - cursor < 8u)
        {
            return invalid(RomFormat::UNIF,
                           CompatibilityReason::UNIF_INVALID_CHUNK,
                           basic_analysis(payload.size));
        }
        const bool prg_chunk =
            payload.data[cursor] == static_cast<std::uint8_t>('P') &&
            payload.data[cursor + 1u] == static_cast<std::uint8_t>('R') &&
            payload.data[cursor + 2u] == static_cast<std::uint8_t>('G') &&
            is_hex_digit(payload.data[cursor + 3u]);
        const std::uint64_t length = little_endian_u32(payload, cursor + 4u);
        cursor += 8u;
        const std::uint64_t remaining = byte_count(payload.size - cursor);
        if (length > remaining)
        {
            return invalid(RomFormat::UNIF,
                           CompatibilityReason::UNIF_INVALID_CHUNK,
                           basic_analysis(payload.size));
        }
        if (prg_chunk && length > 0u)
        {
            has_prg = true;
        }
        cursor += static_cast<std::size_t>(length);
    }
    if (!has_prg)
    {
        return invalid(RomFormat::UNIF,
                       CompatibilityReason::UNIF_MISSING_PRG,
                       basic_analysis(payload.size));
    }

    RomParseResult result;
    result.recognized = true;
    result.format = RomFormat::UNIF;
    result.state = CompatibilityState::UNSUPPORTED;
    result.reason = CompatibilityReason::UNIF_PRODUCT_DISABLED;
    result.analysis = basic_analysis(payload.size);
    return result;
}

} // namespace

RomParseResult parse_rom_payload(ByteView payload) noexcept
{
    if (starts_with_at(payload, 0u, nes_signature))
    {
        return parse_nes(payload);
    }
    if (starts_with_at(payload, 0u, fds_signature))
    {
        return parse_headered_fds(payload);
    }
    if (starts_with_at(payload, 0u, fds_side_signature))
    {
        return parse_headerless_fds(payload);
    }
    if (starts_with_at(payload, 0u, unif_signature))
    {
        return parse_unif(payload);
    }
    return unrecognized(payload.size);
}

} // namespace flynes::catalog
